pkg/objstore/azure/azure.go

```diff
@@ -0,0 +1,292 @@
+package azure
+
+import (
+	"bytes"
+	"context"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	blob "github.com/Azure/azure-storage-blob-go/2018-03-28/azblob"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	yaml "gopkg.in/yaml.v2"
+)
+
+const (
+	opObjectsList  = "ListBucket"
+	opObjectInsert = "PutObject"
+	opObjectGet    = "GetObject"
+	opObjectHead   = "HeadObject"
+	opObjectDelete = "DeleteObject"
+)
+
+// Config Azure storage configuration.
+type Config struct {
+	StorageAccountName string `yaml:"storage-account"`
+	StorageAccountKey  string `yaml:"storage-account-key"`
+	ContainerName      string `yaml:"container"`
+}
+
+// Bucket implements the store.Bucket interface against s3-compatible APIs.
+type Bucket struct {
+	logger       log.Logger
+	containerURL blob.ContainerURL
+	config       *Config
+	opsTotal     *prometheus.CounterVec
+}
+
+// Validate checks to see if any of the s3 config options are set.
+func (conf *Config) Validate() error {
+	if conf.StorageAccountName == "" ||
+		conf.StorageAccountKey == "" {
+		return errors.New("invalid Azure storage configuration")
+	}
+	return nil
+}
+
+// NewBucket returns a new Bucket using the provided Azure config.
+func NewBucket(logger log.Logger, azureConfig []byte, reg prometheus.Registerer, component string) (*Bucket, error) {
+
+	level.Debug(logger).Log("msg", "Creating new Azure bucket connection", "component", component)
+
+	var conf Config
+	if err := yaml.Unmarshal(azureConfig, &conf); err != nil {
+		return nil, err
+	}
+
+	if err := conf.Validate(); err != nil {
+		return nil, err
+	}
+
+	ctx := context.Background()
+
+	container, err := createContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+	if err != nil {
+		ret, ok := err.(blob.StorageError)
+		if !ok {
+			return nil, errors.Wrapf(err, "Azure API return unexpected error: %T\n", err)
+		}
+		if ret.ServiceCode() == "ContainerAlreadyExists" {
+			level.Debug(logger).Log("msg", "Getting connection to existing Azure blob container", "container", conf.ContainerName)
+			container, err = getContainer(ctx, conf.StorageAccountName, conf.StorageAccountKey, conf.ContainerName)
+			if err != nil {
+				return nil, errors.Wrapf(err, "cannot get existing Azure blob container: %s", container)
+			}
+		} else {
+			return nil, errors.Wrapf(err, "error creating Azure blob container: %s", container)
+		}
+	} else {
+		level.Info(logger).Log("msg", "Azure blob container successfully created", "address", container)
+	}
+
+	bkt := &Bucket{
+		logger:       logger,
+		containerURL: container,
+		config:       &conf,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
```

Comment:
Hm.. Do we need this? Seems like all buckets are wrapped with BucketWithMetric in client/factory.go.

Reply:
I would keep it like this since all other providers have it. If we aim to remove it, it must to be done in different PR for all of them.

Reply:
I'm not sure its worth adding it in this PR if we are just going to remove it later very soon.